src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs

```diff
@@ -207,51 +207,104 @@ private void SignalCompletion()
             }
             _completed = true;
 
-            if (_continuation != null || Interlocked.CompareExchange(ref _continuation, ManualResetValueTaskSourceCoreShared.s_sentinel, null) != null)
+            if (_continuation is null && Interlocked.CompareExchange(ref _continuation, ManualResetValueTaskSourceCoreShared.s_sentinel, null) is null)
             {
-                if (_executionContext != null)
+                return;
+            }
+
+            if (_executionContext is null)
+            {
+                if (_capturedContext is null)
+                {
+                    if (RunContinuationsAsynchronously)
+                    {
+                        ThreadPool.UnsafeQueueUserWorkItem(_continuation, _continuationState, preferLocal: true);
+                    }
+                    else
+                    {
+                        _continuation(_continuationState);
+                    }
+
+                    return;
+                }
+
+                InvokeSchedulerContinuation();
+                return;
+            }
+
+            ExecutionContext? currentContext = ExecutionContext.Capture();
+            // Restore the captured ExecutionContext before executing anything.
+            ExecutionContext.Restore(_executionContext);
+
+            if (_capturedContext is null)
+            {
+                if (RunContinuationsAsynchronously)
                 {
-                    ExecutionContext.RunInternal(
-                        _executionContext,
-                        (ref ManualResetValueTaskSourceCore<TResult> s) => s.InvokeContinuation(),
-                        ref this);
+                    ThreadPool.QueueUserWorkItem(_continuation, _continuationState, preferLocal: true);
```

Comment:
This and it looks like `InvokeInlineContinuation()` below could throw, a try/finally may be better for restoring the original context

Reply:
Can QUWI throw?

Reply:
Yea it would allocate and could throw OOM

Reply:
Changed to put restores in finally blocks